[Driver][X86] Add -m[no-]apxf to m_x86_Features_Group

fzou1 opened the pull request:
This is to expose these options to clang-cl on Windows.

fzou1 committed:
[Driver][X86] Add -m[no-]apxf to m_x86_Features_Group

This is to expose these options to clang-cl on Windows.
And add help text for these options.

KanRobert commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6679,8 +6679,13 @@ def mapx_features_EQ : CommaJoined<["-"], "mapx-features=">, Group<m_x86_Feature
+def mno_apxf : Flag<["-"], "mno-apxf">, Alias<mno_apx_features_EQ>,
+               AliasArgs<["egpr","push2pop2","ppx","ndd","ccmp","nf","cf","zu"]>,
+               Group<m_x86_Features_Group>,
+               HelpText<"Disable all features of APX">;
```
The description is not accurate. APX -> APX_F.

BTW,  I suggest to remove the HelpText for m[no-]apxf since other features has no text either, so that they would not appear in `--help`

fzou1 replied:
Updated. Is it expected not to show the feature option in `--help`? Why?

KanRobert replied:
Not clear about that, it's current convention. Probably `-march/mcpu` is preferred.

fzou1 replied:
Okay. Thanks.

fzou1 committed:
Remove help text for m[no_]apxf.

MaskRay commented on clang/test/Driver/cl-x86-flags.c:
```diff
@@ -135,3 +135,53 @@
+// APXF: "-target-feature" "+egpr" "-target-feature" "+push2pop2" "-target-feature" "+ppx" "-target-feature" "+ndd" "-target-feature" "+ccmp" "-target-feature" "+nf" "-target-feature" "+cf" "-target-feature" "+zu"
+// NO-APXF: "-target-feature" "-egpr" "-target-feature" "-push2pop2" "-target-feature" "-ppx" "-target-feature" "-ndd" "-target-feature" "-ccmp" "-target-feature" "-nf" "-target-feature" "-cf" "-target-feature" "-zu"
+
+// RUN: %clang_cl -target x86_64-pc-windows -mapx-features=egpr %s -### -o %t.o 2>&1 | FileCheck -check-prefix=EGPR %s
```
`-target ` has been deprecated around clang 3.4. Use `--target=` in new tests.

Do we really need so many RUN lines?

KanRobert replied:
Should we only test -m[no-]apxf/, -mapx-features are usable for clang-cl? Since other tests are already tested on clang side. @fzou1

fzou1 replied:
Okay. Removed some RUN lines.

MaskRay commented on clang/test/Driver/cl-x86-flags.c:
```diff
@@ -135,3 +135,53 @@
+// NON-APX:      error: unsupported option '-mapx-features=|-mapxf' for target 'i386'
+// NON-APX-NOT:  error: {{.*}} -mapx-features=
+
+// RUN: %clang_cl -target x86_64-pc-windows -mapxf %s -### -o %t.o 2>&1 | FileCheck -check-prefix=APXF %s
```
We don't test the 4 combinations. 

We usually just test `-mno-apxf -mapxf`.

fzou1 replied:
Okay. Updated.

fzou1 committed:
Removed some test lines to save run time for LIT tests

And updated -target option to --target in cl-x86-flags.c and
x86-target-features.c tests as -target option was deprecated.

fzou1 committed:
Removed "-o t.o"

fzou1 commented on clang/test/Driver/cl-x86-flags.c:
```diff
@@ -135,3 +135,53 @@
+
+// RUN: not %clang_cl -target x86_64-pc-windows -mapx-features=egpr,foo,bar %s -### -o %t.o 2>&1 | FileCheck -check-prefix=ERROR %s
+//
+// ERROR: unsupported argument 'foo' to option '-mapx-features='
```
Done.

fzou1 commented:
@KanRobert , any further comments?

fzou1 commented:
> Looks like this breaks tests on macOS: http://45.33.8.238/macm1/107398/step_6.txt
> 
> Please take a look and revert for now if it takes a while to fix.

Thank you. I've reproduced this issue on MacOS and updated PR #141486 to fix it.

rorth commented:
The Solaris/amd64 bot is green again, thanks.

I'd tried to reproduce the failure locally yesterday, but weirdly couldn't: no idea what's the difference between the bot and
my local builds.

fzou1 commented:
> The Solaris/amd64 bot is green again, thanks.

Glad to hear that.
